### PR TITLE
feat: introduce `PhpUnitAttributesFixer`

### PR DESCRIPTION
**Pull request by kubawerlos**
https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7832 and https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7833 are reverted here - it was a (naive) thinking, a better approach is NOT to migrate to attributes that are not supported in a given scope (like `@large` and `@medum` are fine on method level, but `#[Large]` and `#[Medium]` are not allowed ion method level).

### CODE REVIEW

File: src/Fixer/PhpUnit/PhpUnitAttributesFixer.php [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+use PhpCsFixer\Tokenizer\CT;
+use PhpCsFixer\Tokenizer\Token;
+use PhpCsFixer\Tokenizer\Tokens;
+
+final class PhpUnitAttributesFixer extends AbstractPhpUnitFixer
+{
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        return new FixerDefinition(
+            'PHPIUnit attributes must be used over annotations.',
```
Comment (Wirone) [inline src/Fixer/PhpUnit/PhpUnitAttributesFixer.php:33]:
```suggestion
            'PHPUnit attributes must be used over their respective PHPDoc-based annotations.',
```

**Review by Wirone — COMMENTED** [review]
Quick feedback after taking a look 😉. More when it comes out of the draft.

File: src/Fixer/PhpUnit/PhpUnitAttributesFixer.php [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+use PhpCsFixer\Tokenizer\Token;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author Kuba Werłos <werlos@gmail.com>
+ */
+final class PhpUnitAttributesFixer extends AbstractPhpUnitFixer
+{
+    /** @var array<string, string> */
+    private array $fixingMap = [];
```
Comment (mvorisek) [inline src/Fixer/PhpUnit/PhpUnitAttributesFixer.php:36]:
```suggestion
    private array $fixingMap;
```

File: src/Fixer/PhpUnit/PhpUnitAttributesFixer.php [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+        return self::createAttributeTokens($tokens, $index, self::getAttributeNameForAnnotation($annotation));
+    }
+
+    /**
+     * @return list<Token>
+     */
+    private static function fixWithSingleStringValue(Tokens $tokens, int $index, Annotation $annotation): array
+    {
+        Preg::match(
+            sprintf('/@%s\s+(.*\S)(?:\R|\s*\*+\\/$)/', $annotation->getTag()->getName()),
```
Comment (mvorisek) [inline src/Fixer/PhpUnit/PhpUnitAttributesFixer.php:213]: - the regex would be good to be deduplicated
- link to a parsing phpunit source would be good so the parsing can be checked if 1:1 with native phpunit

Comment (kubawerlos) [reply src/Fixer/PhpUnit/PhpUnitAttributesFixer.php:213]: > * the regex would be good to be deduplicated

done

> * link to a parsing phpunit source would be good so the parsing can be checked if 1:1 with native phpunit

I'd rather not, for 2 reasons:
1. it is not a single place:
- https://github.com/sebastianbergmann/phpunit/blob/11.0.3/src/Metadata/Parser/AnnotationParser.php#L63
- https://github.com/sebastianbergmann/phpunit/blob/11.0.3/src/Metadata/Parser/Annotation/DocBlock.php#L41-L45
also, it was made differently in older PHPUnit: https://github.com/sebastianbergmann/phpunit/blob/9.6.16/src/Util/Annotation/DocBlock.php

2. let's focus on providing good tests, so they keep the fixer working properly

**Review by julienfalque — APPROVED** [review]

File: tests/Fixer/PhpUnit/PhpUnitAttributesFixerTest.php [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+            }
+            PHP];
+
+        yield 'do not fix when not supported by attributes' => [<<<'PHP'
+            <?php
+            /**
+             * @covers FooClass::FooMethod
+             * @uses ClassName::methodName
+             */
+            class FooTest extends \PHPUnit\Framework\TestCas {
```
Comment (Wirone) [inline tests/Fixer/PhpUnit/PhpUnitAttributesFixerTest.php:84]: Are the typos (`TestCas`) here intentional?

**Review by Wirone — APPROVED** [review]
Awesome work @kubawerlos 👏. Personally I would drop `handle` from test cases' names because when you consider `--testdox` output it looks weird (e.g. "Fix with handle·After"). Actually I would switch to testdox entirely, but it's a separate story 😉. Anyway, only one minor comment from me, please merge on your own with or without the fix.

File: tests/Fixer/PhpUnit/PhpUnitAttributesFixerTest.php (hunk 4 above)
Comment (kubawerlos) [reply tests/Fixer/PhpUnit/PhpUnitAttributesFixerTest.php]: Those are typos 😄

**Comment by stof** [comment]
it would be great to have an option to add the attributes without removing the annotation (which also requires checking if the attribute is already there to avoid "fixing" it multiple times). When you need to support both PHPUnit 11 (to support the latest PHPUnit version) and PHPUnit 9.6 (to support older PHP versions), the way to go is to define both the attributes (for PHPUnit 10+) and the annotation (used by PHPUnit 9 and ignored in 10+ when the attributes are present).

**Comment by GrahamCampbell** [comment]
Shouldn't this be marked as risky?

**Pull request merged**